tests/integration/utils/kiali_client.go

```diff
@@ -0,0 +1,64 @@
+package utils
+
+import (
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+	"github.com/kiali/kiali/util/httputil"
+)
+
+var kialiURL string
+var kialiToken string
+var kialiCookies []*http.Cookie
+
+func init() {
+	kialiURL = os.Getenv("URL")
+	kialiToken = os.Getenv("TOKEN")
+	if kialiURL == "" || kialiToken == "" {
+		log.Fatalf("URL and TOKEN environment variables are required. Kiali URL in 'https://kiali-hostname' and TOKEN in 'sha256~' format.")
+		return
+	}
+	tokenResult, tokenCookies := GetCookies()
```

Comment:
I am probably chiming in a little late, but I was expecting closer integration into the Kiali codebase.

Personally, I was expecting more to see tests taking the form of these ones, where you don't really need to deploy Kiali to the cluster: https://github.com/kiali/kiali/blob/master/handlers/metrics_test.go.

So, you would do a direct call to the handler bypassing all that auth machinery, which lets you directly test targeted handler, and also would let you mock any 3rd party services and would remove the need to have a full stack.

The need of the full stack, and also the fact that the API is tested through an HTTP client, gives me the idea that this would be closer to the "system tests" side that the "integration tests" side. But "system tests" is what cypress would be covering, isn't it?

I don't know... at the current state, I'm expecting to see a lot of coverage duplicity between these go tests and cypress tests :-/

Reply:
OK, it was my expectations. You have more experience than I about the scope of QE testing if this is going to help on Dev-QE synergy, it's quite OK.

Reply:
As a side comment, I HATE having to write all that mock code that we have to do today. A dev might end up spending more time writing mock code compared to "real" code.

Reply:
> @jmazzitelli  As a side comment, I HATE having to write all that mock code that we have to do today. A dev might end up spending more time writing mock code compared to "real" code.

I do agree. I think sometimes it is neccessary evil. Otherwise, how would you achive modest coverage (i.e. test something different than the happy path)?

> @nrfox These backend api e2e tests would be owned by the dev team since we maintain the kiali api. The effort here is just to port those python tests into golang because the dev team is more familiar with golang than python.

Well... If these go test are going to be owned by the Kiali team, I *do* prefer these to take proper form of integration tests rather than being closer to systems tests :-/

Personally, I envision devs to be doing quite a lot of mocks to achive proper code coverage, while QE-owned tests take care of testing with the full stack.

Just my $0.02. Although, I probably already missed my chance to have any influence here...

Reply:
> As a side comment, I HATE having to write all that mock code that we have to do today. A dev might end up spending more time writing mock code compared to "real" code.

BTW, look at this: https://github.com/dnaeon/go-vcr
It makes me remember... oh... well... :smile: